### PR TITLE
Added characters system

### PR DESCRIPTION
**Pull request by navvyua**
- User can create character
- Admin can accept character application
- Admin can deny character application
- User can edit character application after deny
- Admin can review character after user edits

**Screens**
Character creation
image

Character list
image

Character page
image

Validating characters list
image

Deny modal
image

Denied character in characters list
image

Edit application page
image

Fixed application from admin front of view
image

### CODE REVIEW

File: app/controllers/admin/characters_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class Admin::CharactersController < AdminController
+  before_action :authenticated_only
+  before_action :admin_only
+
+  def index
+    @characters = Character.validating
+  end
+
+  def update
+    Character.find(params[:id]).update!(character_params)
```
Comment (IlyaUmanets) [inline app/controllers/admin/characters_controller.rb]: `Character.find(params[:id])` move this to another method, e.g. `current_character`

File: app/controllers/admin_controller.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class AdminController < ApplicationController
+  private
+
+  def admin_only
```
Comment (IlyaUmanets) [inline app/controllers/admin_controller.rb]: I think we should use this method as `before_action` here.

File: app/controllers/characters_controller.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
+class CharactersController < ApplicationController
+  before_action :current_resource_owner
+  before_action :current_resource, only: [:show, :edit, :update]
+  before_action :only_validated_character, only: :show
+  before_action :only_denied_character, only: [:edit, :update]
+
+  def index
+    @characters = @user.characters.includes(:user)
```
Comment (IlyaUmanets) [inline app/controllers/characters_controller.rb]: do we need `includes(:user)` here ?

File: app/controllers/characters_controller.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    if @character.save
+      redirect_to user_characters_path(@user)
+    else
+      render :new
+    end
+  end
+
+  def update
+    if @character.update(character_params)
+      if @character.validating?
```
Comment (IlyaUmanets) [inline app/controllers/characters_controller.rb]: Please, explain me why do we need this ?

File: app/controllers/characters_controller.rb (hunk 3 above)
Comment (navvyua) [inline app/controllers/characters_controller.rb]: Bullet gem was giving warning when I was loading on index.

File: app/helpers/character_helper.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+module CharacterHelper
+  def format_driving_license(character)
+    character.driving_license ? fa_icon('check', class: 'text-success') : fa_icon('times', class: 'text-danger')
```
Comment (IlyaUmanets) [inline app/helpers/character_helper.rb]: `character.driving_license ? fa_icon_success : fa_icon_danger`

File: app/helpers/character_helper.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  def format_money(value)
+    number_to_currency(value, delimiter: ' ', precision: 0)
+  end
+
+  def validating_characters_count
+    Character.validating.count
+  end
+
+  def character_button(character)
+    if character.denied?
```
Comment (IlyaUmanets) [inline app/helpers/character_helper.rb]: Same here. Use extract method pattern here.

File: app/models/user.rb [hunk 7]
```diff
@@ -3,6 +3,7 @@ class User < ApplicationRecord
 
   has_many :accounts, dependent: :delete_all
   has_many :news
+  has_many :characters
```
Comment (IlyaUmanets) [inline app/models/user.rb]: What about `dependent` ?

File: app/views/characters/_character.html.slim [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+          li.list-group-item
+            b = t('.cash')
+            .pull-right = format_money(character.cash)
+          
+          li.list-group-item
+            b = t('.bank')
+            .pull-right = format_money(character.bank)
+        
+        = character_button(character)
+
```
Comment (IlyaUmanets) [inline app/views/characters/_character.html.slim]: unnecessary line

File: app/controllers/characters_controller.rb (hunk 4 above)
Comment (navvyua) [inline app/controllers/characters_controller.rb]: When character is in validating state user not available to move in character show page, so he will be redirected to index. It happens when user Edit application after denied.

File: app/controllers/admin/characters_controller.rb (hunk 1 above)
Comment (navvyua) [inline app/controllers/admin/characters_controller.rb]: Fixed.

File: app/controllers/admin_controller.rb (hunk 2 above)
Comment (navvyua) [inline app/controllers/admin_controller.rb]: Fixed.

File: app/controllers/characters_controller.rb (hunk 4 above)
Comment (navvyua) [inline app/controllers/characters_controller.rb]: Anyway, fixed.

File: app/helpers/character_helper.rb (hunk 5 above)
Comment (navvyua) [inline app/helpers/character_helper.rb]: Fixed.

File: app/helpers/character_helper.rb (hunk 6 above)
Comment (navvyua) [inline app/helpers/character_helper.rb]: Fixed.

File: app/models/user.rb (hunk 7 above)
Comment (navvyua) [inline app/models/user.rb]: Added delete_all.

File: app/views/characters/_character.html.slim (hunk 8 above)
Comment (navvyua) [inline app/views/characters/_character.html.slim]: Fixed.

**Pull request merged**